trackma/utils.py

```diff
@@ -490,6 +490,53 @@ def get_terminal_size(fd=1):
     return hw
 
 
+# Default 'open folder' function
+def open_folder(engine, show_id, error_callback=None):
+    """
+    Tries to open the folder containing the show of the passed ID.
+    Takes an optional 'error_callback' parameter that lets the invoker specify an error callback method.
+    If no error callback is passed, it defaults to simply raising.
+    """
+
+    try:
+        show_to_open = engine.get_show_info(show_id)
+        filename = engine.get_episode_path(show_to_open)
+
+        with open(os.devnull, 'wb') as DEVNULL:
+            command = []
+
+            match sys.platform:
+                case 'darwin':
+                    command = ["open", os.path.dirname(filename)]
+                case 'win32':
+                    command = ["explorer", os.path.dirname(filename)]
+                case _:
+                    command = ["xdg-open", os.path.dirname(filename)]
+
+            process = subprocess.Popen(command,
+                                       stdout=DEVNULL,
+                                       stderr=subprocess.PIPE,
+                                       universal_newlines=True)
+
+            _, stderr = process.communicate()
+
+            if process.returncode != 0:
+                # Command failed.
+                raise OSError(stderr)
+
+    except OSError as e:
+        if error_callback:
+            error_callback(f'Could not open folder.\n{e}')
+        else:
+            raise e
+
+    except Exception as e:
+        if error_callback:
+            error_callback(e)
+        else:
+            raise e
```

Comment:
This is an anti-pattern. Either the `open_folder` function should handle errors on its own, wrap errors with custom errors or let them bubble, but making this depend on a callback parameter is an unnecessary level of indirection.

With respect to the function providing *some* custom handling for annotating `OSError`s with "Could not open folder", I suggest to wrap only `OSError`s with a custom exception and otherwise let the callers handle the exceptions with their respective UI options.

Reply:
`try: open_foolder(…); catch: …`

Reply:
Regarding the OSError message.
![image](https://github.com/z411/trackma/assets/86712892/23f5a0b1-d6d8-4f2a-a88e-44271ef1e1b2)
I added the string for clarity, but it's not really necessary if you're opposed to it.

If you want to do it through `try: catch:` then you'd have to add that everywhere, where you call open_folder(). Since the function is super simple and its usage doesn't vary in the slightest, using the callback function saves time. I can remove if it and change the wrap the calls in `try:` if you prefer that though

Reply:
Wrapping the error is a valid use case (that I also mentioned in my initial comment). It shouldn't depend on whether an error callback was provided, however, since that's what we have try-catch for. Please just remove the callback parameter and handle thown exceptions at the call site.

(In general, I prefer error handling like it is provided by Rust (or even Golang to some extent), but Python has exceptions and we should use them as they are intended since the rest of the ecosystem is built around them.)

Reply:
Just to make sure I understand what you mean.

Like this?
in `utils.py`
```python
def open_folder(engine, show_id):
    """
    Open the folder containing the show of the passed ID.
    """

    try:
        show_to_open = engine.get_show_info(show_id)
        filename = engine.get_episode_path(show_to_open)

        with open(os.devnull, 'wb') as DEVNULL:
            command = []

            match sys.platform:
                case 'darwin':
                    command = ["open", os.path.dirname(filename)]
                case 'win32':
                    command = ["explorer", os.path.dirname(filename)]
                case _:
                    command = ["xdg-open", os.path.dirname(filename)]

            process = subprocess.Popen(command,
                                       stdout=DEVNULL,
                                       stderr=subprocess.PIPE,
                                       universal_newlines=True)

            _, stderr = process.communicate()

            if process.returncode != 0:
                # Command failed.
                raise OSError(stderr)

    except OSError as e:
        raise OSError(f'Could not open folder: {e}')

    except Exception as e:
        raise RuntimeError(f'An error occurred while trying to open the folder: {e}')
```


in (gtk) `window.py`
```python
    def _open_folder(self, show_id):
        try:
            utils.open_folder(self._engine, show_id)
        except Exception as e:
            self._error_dialog_idle(e)
```

Or do you want me to create a new
```python
class FolderError(TrackmaError):
    pass
```

Reply:
Yes, that is basically what I mean. Either would be fine imo, but a custom `OpenFolderError(TrackmaError)` would be most appropriate since we're interpreting the underlying errors already when re-throwing a wrapped exception. Additionally, you could directly throw a `OpenFolderError` when the subprocess's return code is 0.